compiler/rustc_trait_selection/src/solve/trait_goals.rs

```diff
@@ -120,6 +121,32 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
             return result;
         }
 
+        // Don't call `type_of` on a local TAIT that's in the defining scope,
+        // since that may require calling `typeck` on the same item we're
+        // currently type checking, which will result in a fatal cycle that
+        // ideally we want to avoid, since we can make progress on this goal
+        // via an alias bound or a locally-inferred hidden type instead.
+        //
+        // Also, don't call `type_of` on a TAIT in `Reveal::All` mode, since
+        // we already normalize the self type in
+        // `assemble_candidates_after_normalizing_self_ty`, and we'd
+        // just be registering an identical candidate here.
+        //
+        // Returning `Err(NoSolution)` here is ok in `SolverMode::Coherence`
+        // since we'll always be registering an ambiguous candidate in
+        // `assemble_candidates_after_normalizing_self_ty` due to normalizing
+        // the TAIT.
```

Comment:
I personally don't like this perspective on why we do this. Imo the query cycles are a side effect of something more fundamental:

- `instantiate_constituent_tys_for_auto_trait` for opaque types is a hacky, pseudo normalization-routine which does not check item bounds of the opaque
- if we're able to normalize the opaque in a different way, either because we're able to define it or because we're in `Reveal::All`, using regular normalization is strictly preferable to auto trait leakage

I would like you to rewrite this comment to reflect that, but also fine to wait and chat about it in sync a bit if you don't consider this perspective to be helpful or maybe even correct

Reply:
I'm not particularly attached to any specific comment here, though I think the best way to settle this is for you to just write what you'd like for this comment to say as an inline suggestion.

I'm not sure if we share the exact mental model for why this "hack" is not as much of a hack as it seems, but what I wrote above is how I see it at least from an implementation perspective.

Reply:
the comment I would like to write would also require some general code changes, put that in my notes and will look at it at some later point and open a separate PR